### ISSUE TITLE
Umbrella Tracking Issue For WIP Alpha Feature - Priority and Fairness For API Server Requests 

### ISSUE THREAD

**Issue by aaron-prindle** [kubernetes/kubernetes#76846]
This is to track the changes required for the RequestManagement alpha feature, now that it's [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md) has been approved. A more verbose version of the work can be found in the below [spreadsheet](https://docs.google.com/spreadsheets/d/11v3P9btqqwjlZHpHIarUQoMjse_fU_I_CVCH7QVSXN4/edit?), but the items there are being duplicated here for more transparency.  Feel free to comment and we will update the checklist here if something is not in sync, thanks!

- [Work Tracking Spreadsheet](https://docs.google.com/spreadsheets/d/11v3P9btqqwjlZHpHIarUQoMjse_fU_I_CVCH7QVSXN4/edit?)

# Tasks
## 1) Configuration API Models
- [x] ~~[Submit Draft API Objects for Classification and Prioritization](https://github.com/kubernetes/enhancements/pull/933) @MikeSpreitzer~~
- [x] ~~[[FEATURE BRANCH]: flowcontrol.apiserver.k8s.io initial API models](https://github.com/kubernetes/kubernetes/pull/77048) @yue9944882~~
- [x] ~~Agree on 'PriorityLevelConfiguration' API~~ - ALL
- [x] ~~Agree on 'FlowSchema' API~~ - ALL
- [ ] [registry strategy etc for config API objects](https://github.com/kubernetes/kubernetes/pull/80257) @yue9944882 
- [ ] Discuss proper API abstractions for how to best support future implementation - ALL

## 1a) Handler Skeleton
- [x] ~~[POC concurrency based request handler](https://github.com/yue9944882/inflight-rate-limit) @yue9944882~~
- [x] ~~[Agree on how filter reacts to config changes](https://github.com/kubernetes/enhancements/pull/1098) - ALL~~
- [x] ~~[Started drafting the skeleton of the request management filter](https://github.com/kubernetes/kubernetes/pull/78966) @MikeSpreitzer~~
- [ ] Design Handler Skeleton: Overall Skeleton + Pluggable Interfaces For Classification, Queuing, and Dispatching @MikeSpreitzer
- [ ] Agree on Handler Skeleton Interfaces/Design - ALL
- [ ] Implement Handler Skeleton @MikeSpreitzer

## 1ai) Classification

- [ ] ~~[Implement 'FlowSchema' logic and plug into skeleton](https://github.com/aaron-prindle/kubernetes/pull/1/commits) @yue9944882~~

## 1aii) Queuing & Dispatching - depends somewhat on 2)

- [x] ~~[Shuffle Sharding POC implementation](https://play.golang.org/p/rwtKLj3yGoF) @MikeSpreitzer~~
- [x] ~~[Shuffle Sharding effectiveness evaluation](https://play.golang.org/p/sWvC8nOkPPp) @MikeSpreitzer~~
- [ ] [Shuffle Sharding POC Implementation in tree](https://github.com/kubernetes/kubernetes/pull/80710) @mars1024 
- [ ] Implement dummy queuing & dispatching that will be replaced with 2)
- [ ] ~~[Plug in Fair Queuing Model implementation to skeleton](https://github.com/kubernetes/kubernetes/pull/80786) @aaron-prindle~~
- [ ] Verify that the timeout-handler, which is earlier in the chain and passes stop channel in context, etc works properly

## 1aiii) Add Generation of Default Config Objects

- [ ] Draft Default Config Objects
- [ ] Agree on Default Config Objects and when/how to generate them
- [ ] Implement generation of default config objects

## 2) Fair Queuing Model
- [x] ~~[Implement prototype of modified Fair Queuing for our use case and verify that fairness is provided](https://github.com/aaron-prindle/fq-apiserver/pull/2/commits) @aaron-prindle~~

## 3) Testing

- [ ] Investigate overall test approach // Start a thread around testing approach - ALL
- [ ] Codify the important 'microbenchmark' tests - ALL
- [ ] Codify the important integration broad-use-case tests - ALL
- [ ] Implement 'microbenchmark' test X - UNASSIGNED
- [ ] Implement 'microbenchmark' test Y - UNASSIGNED
- [ ] Implement integration broad-use-case test X - UNASSIGNED
- [ ] Implement integration broad-use-case test Y - UNASSIGNED

## 4) Plumbing & Permissions
- [x] ~~[Add RequestManagement Feature Gate and related flags upstream](https://github.com/kubernetes/kubernetes/pull/76413) @yue9944882~~
- [ ] [Add developers to api-approver to unblock people](https://github.com/kubernetes/kubernetes/pull/79420) @yue9944882
- [ ] [Generalize handler chain constructors to include the new filter] - UNASSIGNED

## 5) Test Validation
Verify that request handler passes all tests from 3) - TBD

/sig api-machinery
/cc @lavalamp @yue9944882 @MikeSpreitzer

**Comment by aaron-prindle** [comment]
Relevant Documents :

- [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md)
- [Simplified Proposal for Alpha](https://github.com/kubernetes/enhancements/pull/933)
- [BucketClasses for Rate-Limiting Tiers](https://docs.google.com/document/d/1BtwFyB6G3JgYOaTxPjQD-tKHXaIYTByKlwlqLE0RUrk/edit#heading=h.kg8s7unxoq91)
- [Bucket-Based In-Flight Request Reserve/Share/Limit](https://docs.google.com/document/d/1xw5m71SKcP4wzeIigWv3wZ3_N9w9G0xHYBJUpxF034w/edit)

Work Graph:
![Work](https://i.ibb.co/tHZMsSn/rsz-request-management-work-graph.png)

**Comment by MikeSpreitzer** [comment]
It's probably time to rebase the feature branch, to pick up the work in `master` before there is any work in the feature branch.

**Comment by aaron-prindle** [comment]
Regarding the Shuffle Sharding POC task, Mike created a POC of an implementation here:
https://play.golang.org/p/rwtKLj3yGoF

The full thread around this is here:
https://github.com/kubernetes/enhancements/pull/933#discussion_r274708574

**Comment by MikeSpreitzer** [comment]
I also created something in the Go Playground to evaluate the effectiveness of shuffle sharding for given settings.  See https://play.golang.org/p/sWvC8nOkPPp .  This ultimately answers the question: for `nHands` elephants, what is the probability (shown there as `probNextCovered`) that a random mouse completely collides with the elephants?

**Comment by yue9944882** [comment]
\+ @mars1024 

tracking testing/benchmark/optimization of shuffle-sharding algorithm

**Comment by mars1024** [comment]
> + @mars1024
> 
> tracking testing/benchmark/optimization of shuffle-sharding algorithm

ref to [https://github.com/mars1024/shuffle-sharding](https://github.com/mars1024/shuffle-sharding) , WIP

**Comment by MikeSpreitzer** [comment]
We need some maintenance of the checklist here.

I updated the spreadsheet in four ways.  I added the completed task of defining the API object types, and the in-progress task of the registry strategy &etc for the config API objects.  I added a task for making the handler chain constructors invoke the new filter.  I added a task for agreeing on how the filter reacts to config changes.

**Comment by aaron-prindle** [comment]
I've synced updated the checklist, syncing with all of the work completed so far and adding the changes mentioned.  Additionally synced some info back into the doc that was in the checklist.  Let me know if anything is missing, I believe both are in sync now.  Thanks.

**Comment by aaron-prindle** [comment]
PR Roundup - 8/14/2018:
https://github.com/kubernetes/kubernetes/pull/81409
https://github.com/kubernetes/kubernetes/pull/81224

https://github.com/kubernetes/kubernetes/pull/81182
https://github.com/kubernetes/kubernetes/pull/81003
https://github.com/kubernetes/kubernetes/pull/80955

https://github.com/kubernetes/kubernetes/pull/80786

https://github.com/kubernetes/enhancements/pull/1200
https://github.com/kubernetes/enhancements/pull/1198

**Comment by MikeSpreitzer** [comment]
I opened #81485 to try to help #81224 along, but am stuck at the integration test failures regading webhooks.

**Comment by MikeSpreitzer** [comment]
FYI, here is how I launch for testing right now:
```
RUNTIME_CONFIG=flowcontrol.apiserver.k8s.io/v1alpha1=true FEATURE_GATES=RequestManagement=true LOG_LEVEL=7 hack/local-up-cluster.sh
```

**Comment by MikeSpreitzer** [comment]
Here are the PRs I see open right now.

#81679 --- more uniform and informative debug logging
#81672 --- add NameAll
#81637 --- add ObjectNamespaces
#81632 --- Stopped rejection of unsecured requests
#81503 --- Fixes missing authenticated group for loopback
#81182 --- More work on shuffle sharding utils
#81003 --- original config controller
#80955 --- Factor PassiveClock out of clock.Clock
#81100 --- flake that is still killling us because we have not rebased after the fix
#79776 --- Compendium of issues with models

https://github.com/kubernetes/enhancements/pull/1198 --- terminology catch-up and ordering considered normal
https://github.com/kubernetes/enhancements/pull/1200 --- one proposal for addressing both default/initialization problems
https://github.com/kubernetes/enhancements/pull/1214 --- another proposal for default/initialization problems plus changes to the designed set of config objects

**Comment by MikeSpreitzer** [comment]
Here is a Go Playground exercise that evaluates an interesting collection of shuffle sharding configurations, printing for each the probability that a given mouse is squished by an elephant for an illustrative collection of numbers of elephants.   See https://play.golang.org/p/s5H-eIpfyc9 , which computes this table:
```
Probability of a given mouse being squished
HandSize,	DeckSize:	1 elephant,		4 elephants,		16 elephants
      12,	      32:	4.428838398950118e-09,	0.11431348830099144,	0.9935089607656024
      10,	      32:	1.550093439632541e-08,	0.0626479840223545,	0.9753101519027554
      10,	      64:	6.601827268370426e-12,	0.00045571320990370776,	0.49999929150089345
       9,	      64:	3.6310049976037345e-11,	0.00045501212304112273,	0.4282314876454858
       8,	      64:	2.25929199850899e-10,	0.0004886697053040446,	0.35935114681123076
       8,	     128:	6.994461389026097e-13,	3.4055790161620863e-06,	0.02746173137155063
       7,	     128:	1.0579122850901972e-11,	6.960839379258192e-06,	0.02406157386340147
       7,	     256:	7.597695465552631e-14,	6.728547142019406e-08,	0.0006709661542533682
       6,	     256:	2.7134626662687968e-12,	2.9516464018476436e-07,	0.0008895654642000348
       6,	     512:	4.116062922897309e-14,	4.982983350480894e-09,	2.26025764343413e-05
       6,	    1024:	6.337324016514285e-16,	8.09060164312957e-11,	4.517408062903668e-07
```

**Comment by MikeSpreitzer** [comment]
Sep 20 morning round-up:

#81834 - rebase the feature branch and merge it into main.  We strongly prefer to proceed from here rather than the current feature branch contents because they are based on such an outdated starting point.

#82196 - extend 81834 for flexibility and bit flips

#81748 - @lavalamp review comments on feature branch

#81637 - Added ObjectNamespaces to PolicyRuleWithSubjects (to establish missing expressiveness)

#81503 - Fixes missing authenticated group for apiserver loopback connection

#79805 - Adaptive retry-after interval on rate limiting to prevent request overload (not part of this feature but improves interaction with it)

#79776 - Apiserver flow-control API model tracker kind/feature

#76742 - Add warning messages to RequestManagement flags. Fix typo

https://github.com/kubernetes/enhancements/issues/1040 - Enhancement main issue

https://github.com/kubernetes/enhancements/pull/1200 - An older attempt to expand KEP's discussion of default behavior issues

https://github.com/kubernetes/enhancements/pull/1214 - Updates KEP to be consistent with #81834

https://github.com/kubernetes/enhancements/pull/1223 - Extends https://github.com/kubernetes/enhancements/pull/1214 consistently with #82196

https://github.com/kubernetes/enhancements/pull/1225 - Add design rationale for original example config objects

The following are hopefully moot because we will proceed from 81834 instead of the feature branch

#81632 - Stopped rejection of unsecured requests (NB: solved a different way in master)

**Comment by MikeSpreitzer** [comment]
We have begun merging and polishing work from the feature branch in pieces.

Already merged:
- #83665 shuffle sharding
- #83671 API objects
- #84706 full validation of non-resource URL prefixes

In progress:
- #84544 fair queuing
- #84708 storage for the API objects
- #84771 refactor priority level config API
- #84900 enable discrimination on target namespace

Still to come:
- the actual filter
  - request handler
  - configuration controller
- further polishing
- agreement on and implementation of "default behaviors"

**Comment by MikeSpreitzer** [comment]
We got a lot done in the past week.  Here are the open PRs now:

- #85321 add server-side printers
- #85319 add request filter and local config controller
- #85268 add cluster initialization

As we do not yet have all the code merged, it is not well tested.
Also some comments in interfaces need fixing.
Unit tests could be better.
Integration and end-to-end tests need to be created.
The user-facing doc is mostly still to be written.
